### ISSUE TITLE
Reproducible builds

### ISSUE THREAD

**Issue by AkihiroSuda** [docker-library/official-images#16044]
Enabling [reproducible builds](https://reproducible-builds.org/) for Docker Official Images will improve traceability of the image origins and will help assessment of supply chain security of the images.

I also talked about this at DockerCon 2023: https://medium.com/nttlabs/dockercon-2023-reproducible-builds-with-buildkit-for-software-supply-chain-security-0e5aedd1aaa7

## Scope of reproduction
The digests of the image manifest blobs (and config blobs and layer blobs) should be reproducible.
<https://github.com/reproducible-containers/diffoci> can be used for testing reproducibility.

SLSA provenance manifest blobs (enabled by default in recent buildx) are not reproducible by design, so the image index digest will *not* be reproducible.

## Reproducing base images

Base images have to be pinned by the sha256 digest for reproduction.

A digest can be embedded in a `FROM` instruction of a Dockerfile.
However, I wonder that image maintainers might not want to update Dockerfiles frequently to ensure the latest base image to be picked up during the upstream build.

In that case, we can just leave `FROM` instructions unpinned, and let reproducers to use the `COVNERT` action of source policies to dynamically replace the image identifier:

https://github.com/moby/buildkit/blob/v0.13.0-beta1/docs/build-repro.md

```json
{
  "action": "CONVERT",
  "selector": {
    "identifier": "docker-image://docker.io/library/alpine:latest"
  },
  "updates": {
    "identifier": "docker-image://docker.io/library/alpine:latest@sha256:4edbd2beb5f78b1014028f4fbb99f3237d9561100b6881aabbf5acce2c4f9454"
  }
}
```

The digest to be used for the `CONVERT` action is recored in the SLSA provenance.
We need to update buildx CLI to automatically generate `CONVERT` actions from SLSA provenances.
(See the `buildx CLI` section below)

### Debian and Ubuntu

Debian and Ubuntu keep old packages on http://snapshot.debian.org and http://snapshot.ubuntu.com.
I wrote a script to rewrite `/etc/apt/sources.list` to use those snapshot servers:
<https://github.com/reproducible-containers/repro-sources-list.sh/blob/master/repro-sources-list.sh>
The snapshot timestamp can be supplied via `$SOURCE_DATE_EPOCH`.

These snapshot servers are quite slow and sometimes flaky (especially for Debian), so, probably the snapshot servers
shouldn't be used for the upstream builds.
Reproducers may provide `--build-arg` to opt-in to use the apt snapshots.

For <https://github.com/docker-library/gcc>, the change will look like as follows:

```patch
diff --git a/Dockerfile.template b/Dockerfile.template
index cc90d0e..7474df5 100644
--- a/Dockerfile.template
+++ b/Dockerfile.template
@@ -31,8 +31,21 @@ ENV GCC_MIRRORS \
 ENV GCC_VERSION {{ .version }}
 # Docker EOL: {{ .eol }}
 
+ADD --chmod=0755 --checksum=sha256:4c97fd03a3b181996b1473f3a99b69a1efc6ecaf2b4ede061b6bd60a96b9325a \
+       https://raw.githubusercontent.com/reproducible-containers/repro-sources-list.sh/v0.1.0/repro-sources-list.sh \
+       /usr/local/bin/repro-sources-list.sh
+
+# REPRO reproduces apt-get package versions
+ARG REPRO=0
+# SOURCE_DATE_EPOCH is consumed by repro-sources-list.sh
+ARG SOURCE_DATE_EPOCH
+
 RUN set -ex; \
        \
+       if [ "$REPRO" = "1" ]; then \
+               cp -a /etc/apt /etc/apt.bak; \
+               repro-sources-list.sh; \
+       fi; \
        savedAptMark="$(apt-mark showmanual)"; \
        apt-get update; \
        apt-get install -y --no-install-recommends \
@@ -41,6 +54,10 @@ RUN set -ex; \
                gnupg \
        ; \
        rm -r /var/lib/apt/lists/*; \
+       if [ "$REPRO" = "1" ]; then \
+               rm -rf /etc/apt; \
+               mv /etc/apt.bak /etc/apt; \
+       fi; \
        \
        _fetch() { \
                local fetch="$1"; shift; \
```

### Alpine
Reproducing apk packages is still challenging, as Alpine does not have snapshot servers.

The long-term plan is to capture apk packages on building and attach them to the image as artifacts:
- https://github.com/moby/buildkit/pull/4286

## Reproducing file timestamps

BuildKit v0.13 (still beta at the time of writing this) supports rewriting the timestamps of the files inside image layers to use `$SOURCE_DATE_EPOCH`.
```
--output type=image,name=docker.io/username/image,push=true,rewrite-timestamp=true
```

https://github.com/moby/buildkit/blob/v0.13.0-beta1/docs/build-repro.md#source_date_epoch

## Reproducing file contents

Some dockerfiles will need extra work for reproducing file contents.
e.g.,  sorting arrays, removing randomized mktemp, ...

- https://reproducible-builds.org/docs/stable-outputs/
- https://reproducible-builds.org/docs/randomness/

## Buildx CLI

Buildx CLI should be updated to allow attesting reproducibility with a few commands.
Notably, `buildx build` should have a flag like `--repro from=gcc@sha256@...` to import build args and base image digests from an SLSA provenance:
```console
$ # "none://" is a filler for the build context arg
$ docker buildx build \
  --load \
  -t gcc:local \
  --repro from=gcc@sha256:f97e2719cd5138c932a814ca43f3ca7b33fde866e182e7d76d8391ec0b05091f \
  none://
...
[amd64] Using SLSA provenance sha256:7ecde97c24ea34e1409caf6e91123690fa62d1465ad08f638ebbd75dd381f08f
[amd64] Importing Dockerfile blob embedded in the provenance
[amd64] Importing build context https://github.com/docker-library/gcc.git#af458ec8254ef7ca3344f12631e2356b20b4a7f1:13
[amd64] Importing build-arg SOURCE_DATE_EPOCH=1690467916
[amd64] Importing buildpack-deps:bookworm from docker-image://buildpack-deps:bookworm@sha256:bccdd9ebd8dbbb95d41bb5d9de3f654f8cd03b57d65d090ac330d106c87d7ed
...

$ diffoci diff gcc@sha256:f97e2719cd5138c932a814ca43f3ca7b33fde866e182e7d76d8391ec0b05091f gcc:local
...
```

## CI
We will also need to have a CI to periodically attest reproducibility with the proposed CLI above.

**Comment by sudo-bmitch** [comment]
> SLSA provenance manifest blobs (enabled by default in recent buildx) are not reproducible by design, so the image index digest will not be reproducible.

The index could be reproducible if the artifacts were associated using the upcoming OCI subject/referrers API rather than injecting it directly in the index.

> A digest can be embedded in a FROM instruction of a Dockerfile.
> However, I wonder that image maintainers might not want to update Dockerfiles frequently to ensure the latest base image to be picked up during the upstream build.
> 
> In that case, we can just leave FROM instructions unpinned, and let reproducers to use the COVNERT action of source policies to dynamically replace the image identifier

The push from groups like Scorecard is to embed the digest directly in the Dockerfile and maintain the pin with a tool like renovate or dependabot. This gives reproducibility with controlled updates.

> We will also need to have a CI to periodically attest reproducibility with the proposed CLI above.

I'd love to see work done on a user tool that can take any built image, or git commit, and verify the output. We can assist this work by annotating the images with details like the git commit used to build them and the `SOURCE_DATE_EPOC` value. This could allow an independent 3rd party verification of images, and signatures or attestations pushed to their repository as detached artifacts (artifacts referencing an image that is not in that repository). Users could then add a policy on their side to require a 3rd party attestation from a list of trusted rebuilders.

**Comment by AkihiroSuda** [comment]
> The index could be reproducible if the artifacts were associated using the upcoming OCI subject/referrers API rather than injecting it directly in the index.

Yes, but I guess adoption of the spec v1.1 in DOI is likely to take years, as user's site-local mirrors are mostly not ready for v1.1 yet

> The push from groups like Scorecard is to embed the digest directly in the Dockerfile and maintain the pin with a tool like renovate or dependabot. This gives reproducibility with controlled updates.

Yes, but I wonder that maintainers might not want to see tons of bump-up PRs to happen every day.

**Comment by sudo-bmitch** [comment]
> > The index could be reproducible if the artifacts were associated using the upcoming OCI subject/referrers API rather than injecting it directly in the index.
> 
> Yes, but I guess adoption of the spec v1.1 in DOI is likely to take years, as user's site-local mirrors are mostly not ready for v1.1 yet

OCI 1.0 conformant mirrors already support it by use of a fallback tag. It's similar to how sigstore implements their attestations today.

> > The push from groups like Scorecard is to embed the digest directly in the Dockerfile and maintain the pin with a tool like renovate or dependabot. This gives reproducibility with controlled updates.
> 
> Yes, but I wonder that maintainers might not want to see tons of bump-up PRs to happen every day.

There's a constant bump-up already. One is invisible and hurts reproducibility, while the other is documented with a Git commit.

**Comment by AkihiroSuda** [comment]
The first PR:
- https://github.com/docker-library/httpd/pull/248

**Comment by tianon** [comment]
I don't have the bandwidth to respond to all of this right now, so I'll focus on the bit that concerns me the most: the service provided at `snapshot.debian.org` is not currently well-maintained or well-staffed (hence the speed issues, but it's a much deeper problem), so I would be very uncomfortable intentionally adding to them the load of all our image builds/rebuilds.

At a high level, I think the best place to start here (especially as it would be the least disruptive) is making useful/interesting _layers_ reproducible.  For example, https://github.com/docker-library/golang/commit/46f40bd7c5706e873f5177719159849baa95b275 is an old PoC that @yosifkit worked on which provides that for the Go-providing layer of the Go images that I've been planning to revisit in the near future.

**Comment by AkihiroSuda** [comment]
> I don't have the bandwidth to respond to all of this right now, so I'll focus on the bit that concerns me the most: the service provided at `snapshot.debian.org` is not currently well-maintained or well-staffed (hence the speed issues, but it's a much deeper problem), so I would be very uncomfortable intentionally adding to them the load of all our image builds/rebuilds.

This proposal does NOT enable `snapshot.debian.org` for the upstream builds.
Third-party reproducers may opt-in to use `snapshot.debian.org` by providing `--secret id=enable-repro-sources-list,source=/dev/null` to repro the packages that were on the regular debian.org at the time of the `SOURCE_DATE_EPOCH`.
(This secret is not really "secret" but made as a secret to avoid affecting the `history` object in OCI Image Config blob)

Could you take a look at the PR again ?
- https://github.com/docker-library/httpd/pull/248

> At a high level, I think the best place to start here (especially as it would be the least disruptive) is making useful/interesting _layers_ reproducible. For example, [docker-library/golang@46f40bd](https://github.com/docker-library/golang/commit/46f40bd7c5706e873f5177719159849baa95b275) is an old PoC that @yosifkit worked on which provides that for the Go-providing layer of the Go images that I've been planning to revisit in the near future.

In the case of `httpd` image, the useful/interesting layer is the layer of the binaries such as `/usr/local/bin/httpd`.
To reproduce the binaries we have to reproduce at the compiler packages.

**Comment by AkihiroSuda** [comment]
BTW `snapshot.ubuntu.com` hosted on Azure seems quite fast enough to adopt as the default.
https://ubuntu.com/blog/ubuntu-snapshots-on-azure-ensuring-predictability-and-consistency-in-cloud-deployments

Would it be acceptable to add Ubuntu variants to DOI?
This will be also beneficial for Ubuntu users who are not interested in reproducible builds

**Comment by AkihiroSuda** [comment]
@tianon Could you take a look? 🙏

**Comment by codethief** [comment]
I'm only an interested bystander but I've long been looking for reproducible images, so that, using them as base images, I can create reproducible images for my projects, too.

@AkihiroSuda Would your proposal also include publishing those images on the daily, tagged by date (i.e. corresponding to the snapshot package sources), like @tianon currently [does for Debian on a ~monthly basis](https://hub.docker.com/r/debian/snapshot).

This would be incredibly useful downstream as one could then simply bump the date tag of the base image on the daily to upgrade dependencies & protect against vulnerabilities while still preserving reproducibility.

**Comment by AkihiroSuda** [comment]
> @AkihiroSuda Would your proposal also include publishing those images on the daily, tagged by date (i.e. corresponding to the snapshot package sources), like @tianon [currently does for Debian on a ~monthly basis](https://hub.docker.com/r/debian/snapshot)?

This is an orthogonal topic.

**Comment by jan-kiszka** [comment]
FWIW, I've went through the not-yet-fully-smooth process of making the kas build containers for Yocto and Isar reproducible (https://github.com/siemens/kas/commits/next). Our containers as based on Debian, and it would be great to see our dependencies reproducible as well.

BTW, I had some fun with understanding differences due to cached layers that are apparently not rewritten timestamp-wise. Is that a known issue of BuildKit? I'm now refraining from caching layers persistently on GH and from using the cache completely (unfortunately) when validating locally.

**Comment by AkihiroSuda** [comment]
> BTW, I had some fun with understanding differences due to cached layers that are apparently not rewritten timestamp-wise.

Could you open an issue in https://github.com/moby/buildkit/issues ?

**Comment by jan-kiszka** [comment]
Done: https://github.com/moby/buildkit/issues/4748

**Comment by codethief** [comment]
@tianon 

> the service provided at snapshot.debian.org is not currently well-maintained or well-staffed (hence the speed issues, but it's a much deeper problem), so I would be very uncomfortable intentionally adding to them the load of all our image builds/rebuilds.

I just came across http://snapshot-cloudflare.debian.org . Does your comment also apply to that host (mirror?) or could that one be used more safely for the purposes of image (re-)builds?

**Comment by tianon** [comment]
Yes, I do not think that causes much (if any) less load on the upstream infrastructure given that I do not believe it caches the heavy queries, but cannot be certain as it does not seem to be officially documented or even discussed anywhere public.

There have been recent efforts within the project to reinvigorate support for the snapshot service (see https://lists.debian.org/debian-snapshot/2024/03/msg00000.html for the most recent posted notes) but they are going to take time.

**Comment by AkihiroSuda** [comment]
> http://snapshot-cloudflare.debian.org

This seems still slow.

My current suggestion is to just build the upstream DOI images without pinning, and let downstream reproducers hook the Dockerfile to use `snapshot[-cloudflare].debian.org` or whatever they like.
- https://github.com/docker-library/httpd/pull/250
- https://github.com/moby/buildkit/pull/4669

**Comment by AkihiroSuda** [comment]
Here is the first batch of the PRs:
- https://github.com/docker-library/httpd/pull/250
- https://github.com/tianon/docker-bash/pull/38
- https://github.com/docker-library/haproxy/pull/229
- https://github.com/docker-library/memcached/pull/96
- https://github.com/docker-library/ruby/pull/455


Is there anything left I have to do to get these PRs merged?

I saw a comment about `ARG SOURCE_DATE_EPOCH`, but didn't get what I have to do:
- https://github.com/tianon/docker-bash/pull/38#discussion_r1586880059

**Comment by tianon** [comment]
I'm sorry, this is still on my TODO list, but it is admittedly not a very high priority at the current time.

**Comment by AkihiroSuda** [comment]
> I'm sorry, this is still on my TODO list, but it is admittedly not a very high priority at the current time.

@tianon I appreciate your hard work, and I know you have been very busy, but could you allocate one minute to help understanding your comment about `ARG SOURCE_DATE_EPOCH`? https://github.com/tianon/docker-bash/pull/38#discussion_r1586880059 🙏 

I understand that reviewing and merging PRs may take a longer time, but I want to make sure that we are on the same direction.

**Comment by AkihiroSuda** [comment]
Is there anything I can do to keep this actionable? 🙏

**Comment by AkihiroSuda** [comment]
Also submitted a merge request to add a guide to <https://reproducible-builds.org/> :

- https://salsa.debian.org/reproducible-builds/reproducible-website/-/merge_requests/124

**Comment by fmoessbauer** [comment]
Unfortunately all images since `20240531T083821Z` are no longer reproducible since the upstream mirror snapshots-cloudflare.d.o did not receive updates anymore. This is especially problematic, as the auto-redirect to latest snapshot logic hides this.

Xref: https://github.com/reproducible-containers/repro-sources-list.sh/issues/17

**Comment by AkihiroSuda** [comment]
> Unfortunately all debian images since `20240531T083821Z` are no longer reproducible since the upstream mirror snapshots-cloudflare.d.o did not receive updates anymore. This is especially problematic, as the auto-redirect to latest snapshot logic hides this.
> 
> Xref: [reproducible-containers/repro-sources-list.sh#17](https://github.com/reproducible-containers/repro-sources-list.sh/issues/17)

Looks like we should just use https://snapshot.debian.org/ now

**Comment by fmoessbauer** [comment]
> Looks like we should just use https://snapshot.debian.org/ now
> 

The [ML thread](https://lists.debian.org/debian-snapshot/2024/06/msg00014.html) suggests to use `https://snapshot-mlm-01.debian.org/`, but I don't trust these being long-term stable either. On snapshot.d.o we are rate-limited, making it more or less impractical for real-world use cases.

Anyways, all images since `20240531T083821Z` are lost.

**Comment by AkihiroSuda** [comment]
> On snapshot.d.o we are rate-limited, making it more or less impractical for real-world use cases.

dpkgs can be also cached to an OCI registry or whatever (without altering Dockerfile), if this PR can be merged:
- https://github.com/moby/buildkit/pull/4669

Still I don't know what is the blocker to get this one (and other PRs) merged, though.

**Comment by tianon** [comment]
As far as I have seen, the folks maintaining snapshot do not officially maintain (nor recommend for long-term usage) _any_ hostname/URL other than the canonical snapshot.debian.org, which as of a few weeks ago (see https://lists.debian.org/debian-snapshot/2024/07/msg00000.html for example), includes the new snapshot-mlm-01.debian.org server in the official rotation:

```console
$ dig snapshot-mlm-01.debian.org +short
185.213.153.170
$ dig snapshot.debian.org +short
185.17.185.185
185.213.153.170
```

As I noted in https://github.com/docker-library/official-images/issues/16044#issuecomment-2043589921 _and_ subsequently confirmed via IRC with the folks currently maintaining the snapshot service, I don't know what that snapshot-cloudflare URL is/was, but best guess is a partially implemented PoC that was never completed (and definitely never official advertised/recommended anywhere I'm aware of or can find, even searching old mailing list archives).

**Comment by AkihiroSuda** [comment]
Right, only the canonical `snapshot.debian.org` should be used for reproductions.

Anyway, this only matters for third-parties who want to verify reproducibility of upstream images ( with a dockerfile hook proposed in https://github.com/moby/buildkit/pull/4669 ).

The PRs for the upstream DOI builds do not need to use `snapshot.debian.org` or something similar, so they should be ready to merge as-is:
- https://github.com/docker-library/httpd/pull/250
- https://github.com/tianon/docker-bash/pull/38
- https://github.com/docker-library/haproxy/pull/229
- https://github.com/docker-library/memcached/pull/96
- https://github.com/docker-library/ruby/pull/455

**Comment by AkihiroSuda** [comment]
Opened a PR to set `rewrite-timestamp=true`:
- https://github.com/docker-library/meta-scripts/pull/70

**Comment by AkihiroSuda** [comment]
What's the current blocker for the PRs (linked in the OP)?

**Comment by codethief** [comment]
Looks like snapshot.debian.org recently(?) introduced relatively strict rate limiting. Or at least our pipelines suddenly started failing this morning and our organization is fairly small (a few dozen Docker builds per day and most builds are cached, anyway).

**Comment by AkihiroSuda** [comment]
> snapshot.debian.org

Yes, this server is quite slow and unstable.
This should be only used in reproducers' builds, not in the upstream builds

**Comment by fmoessbauer** [comment]
> Yes, this server is quite slow and unstable. This should be only used in reproducers' builds, not in the upstream builds

Well... If your products should use a stable baseline as well, you also have to use it there. I'm currently working together with the people behind snapshot.d.o to improve the situation. For details, see:

- https://lists.debian.org/debian-snapshot/2024/11/msg00004.html
- https://salsa.debian.org/apt-team/apt/-/merge_requests/383

**Comment by AkihiroSuda** [comment]
Closing as there seems no interest
